### PR TITLE
Add ability to configure Redis Ring (sharded) client instead of single server-client.

### PR DESCRIPTION
**Pull request by vadimberezniker**
The Ring client does not support all the same options that the simple
client does which means we can't create a ring client everywhere.
Notably we support unix sockets in redis target which are not supported
by the ring client.

The go-redis library also does not have a common option type that works
for both the simple and ring clients. It has a a UniversalOptions type
but it works for everything EXCEPT the Ring client so we introduce our
own in redisutil.

---

**Version bump**: Minor

### CODE REVIEW

File: server/config/config.go [hunk 1]
```diff
@@ -251,21 +253,29 @@ type SSLConfig struct {
 	DefaultHost      string   `yaml:"default_host" usage:"Host name to use for ACME generated cert if TLS request does not contain SNI."`
 }
 
+type ShardedRedisConfig struct {
+	Shards   []string `yaml:"shards" usage:"Ordered list of Redis shard addresses."`
+	Username string   `yaml:"username" usage:"Redis username"`
+	Password string   `yaml:"password" usage:"Redis password"`
```
Comment (bduffany) [inline server/config/config.go:259]: would it ever make sense for shards to have their own unique passwords/usernames? (wondering if ShardedRedisConfig could instead be replaced with a list of targets)

Comment (vadimberezniker) [reply server/config/config.go:259]: The redis ring client in go-redis does not support specifying different usernames/passwords for different shards.

File: server/config/config.go [hunk 2]
```diff
@@ -34,37 +34,38 @@ type generalConfig struct {
...
+	GlobalFilterEnabled       bool               `yaml:"global_filter_enabled" usage:"If set, the global filter will be enabled in the UI."`
+	TestGridV2Enabled         bool               `yaml:"test_grid_v2_enabled" usage:"Whether to enable test grid V2"`
+	UsageEnabled              bool               `yaml:"usage_enabled" usage:"If set, the usage page will be enabled in the UI."`
+	UsageStartDate            string             `yaml:"usage_start_date" usage:"If set, usage data will only be viewable on or after this timestamp. Specified in RFC3339 format, like 2021-10-01T00:00:00Z"`
+	UsageTrackingEnabled      bool               `yaml:"usage_tracking_enabled" usage:"If set, enable usage data collection."`
+	DefaultRedisTarget        string             `yaml:"default_redis_target" usage:"A Redis target for storing remote shared state. To ease migration, the redis target from the remote execution config will be used if this value is not specified."`
+	DefaultShardedRedis       ShardedRedisConfig `yaml:"default_sharded_redis" usage:"Configuration for storing ephemeral state across multiple Redis instances. Mutually exclusive with default_redis_target."`
```
Comment (bduffany) [inline server/config/config.go:67]: nit: wdyt about calling this `default_redis_config`? (same for other config fields)

**Review by bduffany — APPROVED** [review]
LGTM, very clean

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: server/config/config.go (hunk 2 above)
Comment (vadimberezniker) [reply server/config/config.go:67]: I think we should keep the current name to make it clear that this is for setting up sharding.
I expect almost every on-prem customer to keep running BB using a non-sharded Redis setup and hopefully this naming will make it clearer that they shouldn't touch the sharded config.

**Pull request merged**